File: src/analysis/tonal.c
#include "algorithm.h"

#include "sounds/analysis.h"

#include <stdlib.h>

typedef struct TonalAlgorithm {
    SoundWaveletAnalyzer *wavelet;
    float *samples;
    uint64_t column_samples;
    uint64_t analyzed_samples;
} TonalAlgorithm;

static void tonal_destroy(void *state) {
    TonalAlgorithm *algorithm = state;

    if (!algorithm) {
        return;
    }

    sound_wavelet_analyzer_destroy(algorithm->wavelet);
    free(algorithm->samples);
    free(algorithm);
}

static double tonal_min_frequency(const void *state) {
    const TonalAlgorithm *algorithm = state;
    return sound_wavelet_analyzer_min_frequency(algorithm->wavelet);
}

static double tonal_max_frequency(const void *state) {
    const TonalAlgorithm *algorithm = state;
    return sound_wavelet_analyzer_max_frequency(algorithm->wavelet);
}

static void tonal_toggle_sst(void *state) {
    TonalAlgorithm *algorithm = state;
    bool enabled = !sound_wavelet_analyzer_synchrosqueezed(algorithm->wavelet);

    sound_wavelet_analyzer_set_synchrosqueezed(algorithm->wavelet, enabled);
}

static bool tonal_sst_enabled(const void *state) {
    const TonalAlgorithm *algorithm = state;
    return sound_wavelet_analyzer_synchrosqueezed(algorithm->wavelet);
}

static bool tonal_push_column(
    TonalAlgorithm *algorithm,
    const SoundAnalysisInput *input,
    bool *pushed,
    SoundError *error
) {
    *pushed = false;

    uint64_t read_count = sound_ring_buffer_read_ending_at(
        input->ring,
        algorithm->analyzed_samples + algorithm->column_samples,
        algorithm->samples,
        algorithm->column_samples
    );

    if (read_count != algorithm->column_samples) {
        algorithm->analyzed_samples = input->written_samples;
        return true;
    }

    if (!sound_wavelet_analyzer_push(
            algorithm->wavelet,
            algorithm->samples,
            read_count,
            error
        )) {
        return false;
    }

    algorithm->analyzed_samples += read_count;
    *pushed = true;
    return true;
}

static bool tonal_update(
    void *state,
    const SoundAnalysisInput *input,
    SoundAnalysisOutput *output,
    bool emit,
    SoundError *error
) {
    TonalAlgorithm *algorithm = state;

    if (input->written_samples > algorithm->analyzed_samples + input->ring_capacity) {
        algorithm->analyzed_samples = input->written_samples - input->ring_capacity;
    }

    uint64_t pending =
        (input->written_samples - algorithm->analyzed_samples) / algorithm->column_samples;
    if (pending > input->column_limit) {
        pending = input->column_limit;
    }

    for (uint64_t column = 0; column < pending; ++column) {
        bool pushed = false;
        if (!tonal_push_column(algorithm, input, &pushed, error)) {
            return false;
        }

        if (!pushed) {
            break;
        }

        if (!emit) {
            continue;
        }

        float *dbfs_rows = NULL;
        if (!sound_analysis_output_append(output, &dbfs_rows, error)) {
            return false;
        }

        if (!sound_wavelet_analyzer_snapshot_db(
                algorithm->wavelet,
                dbfs_rows,
                input->row_count,
                error
            )) {
            return false;
        }
    }

    return true;
}

static const SoundAnalysisAlgorithmOps tonal_ops = {
    .destroy = tonal_destroy,
    .min_frequency = tonal_min_frequency,
    .max_frequency = tonal_max_frequency,
    .reset = NULL,
    .toggle_sst = tonal_toggle_sst,
    .sst_enabled = tonal_sst_enabled,
    .update = tonal_update,
};

bool sound_tonal_algorithm_create(
    double sample_rate,
    uint64_t column_samples,
    SoundAnalysisAlgorithm *algorithm,
    SoundError *error
) {
    *algorithm = (SoundAnalysisAlgorithm){0};

    TonalAlgorithm *state = calloc(1, sizeof(*state));
    if (!state) {
        sound_error_set(error, "could not allocate tonal algorithm");
        return false;
    }

    state->column_samples = column_samples;
    state->samples = malloc(sizeof(float) * (size_t)column_samples);

    if (!state->samples) {
        tonal_destroy(state);
        sound_error_set(error, "could not allocate tonal sample buffer");
        return false;
    }

    if (!sound_wavelet_analyzer_create(sample_rate, &state->wavelet, error)) {
        tonal_destroy(state);
        return false;
    }

    *algorithm = (SoundAnalysisAlgorithm){
        .ops = &tonal_ops,
        .state = state,
        .mode = SOUND_APP_MODE_TONAL,
    };
    return true;
}

File: src/analysis/tonal.zig
//! Streaming tonal wavelet mode with optional synchrosqueezing.

const std = @import("std");
const wavelet = @import("wavelet.zig");
const ring_buffer = @import("../audio/ring_buffer.zig");
const live_columns = @import("live_columns.zig");

const columns_per_update = 16;
const backlog_skip_columns = 64;

pub const Algorithm = struct {
    allocator: std.mem.Allocator,
    wavelet_analyzer: wavelet.Analyzer,
    samples: []f32,
    column_samples: u64,
    analyzed_samples: u64,

    pub fn init(
        allocator: std.mem.Allocator,
        sample_rate: f64,
        column_samples: u64,
    ) !Algorithm {
        const samples = try allocator.alloc(f32, @intCast(column_samples));
        errdefer allocator.free(samples);

        var analyzer = try wavelet.Analyzer.init(allocator, sample_rate);
        errdefer analyzer.deinit();

        return .{
            .allocator = allocator,
            .wavelet_analyzer = analyzer,
            .samples = samples,
            .column_samples = column_samples,
            .analyzed_samples = 0,
        };
    }

    pub fn deinit(self: *Algorithm) void {
        self.wavelet_analyzer.deinit();
        self.allocator.free(self.samples);
        self.* = undefined;
    }

    pub fn minFrequency(self: *const Algorithm) f64 {
        return self.wavelet_analyzer.minFrequency();
    }

    pub fn maxFrequency(self: *const Algorithm) f64 {
        return self.wavelet_analyzer.maxFrequency();
    }

    pub fn setFrequencyRange(self: *Algorithm, min_hz: f64, max_hz: f64) !void {
        try self.wavelet_analyzer.setFrequencyRange(min_hz, max_hz);
    }

    pub fn toggleSst(self: *Algorithm) void {
        self.wavelet_analyzer.setSynchrosqueezed(!self.wavelet_analyzer.isSynchrosqueezed());
    }

    pub fn sstEnabled(self: *const Algorithm) bool {
        return self.wavelet_analyzer.isSynchrosqueezed();
    }

    pub fn reset(self: *Algorithm, written_samples: u64) void {
        self.analyzed_samples = written_samples;
    }

    pub fn update(
        self: *Algorithm,
        ring: *const ring_buffer.RingBuffer,
        written_samples: u64,
        ring_capacity: u64,
        row_count: usize,
        column_limit: usize,
        columns: []f32,
        column_count: *usize,
    ) !void {
        if (row_count == 0 or column_limit == 0) return;

        // Timeline resets can move the write cursor behind this mode's last read.
        if (written_samples < self.analyzed_samples) self.analyzed_samples = written_samples;

        if (written_samples > self.analyzed_samples + ring_capacity) {
            self.analyzed_samples = written_samples - ring_capacity;
        }

        var pending = (written_samples - self.analyzed_samples) / self.column_samples;
        // Keep the frame loop responsive: analyze a small batch per frame
        // (steady state is ~4 columns) and, after a real stall, drop the
        // backlog and restart from the newest window like a timeline reset.
        if (pending > backlog_skip_columns) {
            self.analyzed_samples = written_samples - backlog_skip_columns * self.column_samples;
            pending = backlog_skip_columns;
        }
        pending = @min(pending, @as(u64, @intCast(@min(column_limit, columns_per_update))));

        for (0..@intCast(pending)) |_| {
            const read_count = ring.readEndingAt(
                self.analyzed_samples + self.column_samples,
                self.samples,
            );

            if (read_count != self.samples.len) {
                self.analyzed_samples = written_samples;
                break;
            }

            try self.wavelet_analyzer.push(self.samples);
            self.analyzed_samples += read_count;

            const dbfs_rows = try live_columns.append(columns, row_count, column_limit, column_count);
            try self.wavelet_analyzer.snapshotDb(dbfs_rows);
        }
    }
};
